### PR TITLE
add item_pre_anvil event

### PR DESCRIPTION
**Pull request by Lildirt**
Kind of flying blind with creating a new event. It's been many years since I've written one, in an extension. :joy:
Let me know if there's any validation you'd like for me to do on my end.

Example payload (pretty printed):
```
event_type: item_pre_anvil
first_item: {meta: {damage: 0, display: null, enchants: {silk_touch: {elevel: 1}, unbreaking: {elevel: 3}}, flags: {}, lore: null, model: null, modifiers: null, repair: 0, tags: null, unbreakable: false}, name: DIAMOND_PICKAXE, qty: 1}
item_repair_cost: 10
level_repair_cost: 10
macrotype: inventory
max_repair_cost: 40
result: {meta: {damage: 0, display: null, enchants: {efficiency: {elevel: 4}, silk_touch: {elevel: 1}, unbreaking: {elevel: 3}}, flags: {}, lore: null, model: null, modifiers: null, repair: 1, tags: null, unbreakable: false}, name: DIAMOND_PICKAXE, qty: 1}
second_item: {meta: {damage: 0, display: null, enchants: {efficiency: {elevel: 4}, unbreaking: {elevel: 3}}, flags: {}, lore: null, model: null, modifiers: null, repair: 0, tags: null, unbreakable: false}, name: DIAMOND_PICKAXE, qty: 1}
viewers: {Lildirt}
```

### CODE REVIEW

File: src/main/java/com/laytonsmith/abstraction/bukkit/BukkitMCAnvilInventory.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	@Override
+	public int getRepairCost() {
+		return ai.getRepairCost();
+	}
+
+	@Override
+	public int getRepairCostAmount() {
+		return ai.getRepairCost();
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/abstraction/bukkit/BukkitMCAnvilInventory.java:43]: I believe that this should call `getRepairCostAmount()` instead.

File: src/main/java/com/laytonsmith/abstraction/events/MCPrepareAnvilEvent.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.laytonsmith.abstraction.events;
+
+import com.laytonsmith.abstraction.MCHumanEntity;
+import com.laytonsmith.abstraction.MCInventory;
+import com.laytonsmith.abstraction.MCItemStack;
+
+import java.util.List;
+
+public interface MCPrepareAnvilEvent extends MCInventoryEvent {
+	@Override
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/abstraction/events/MCPrepareAnvilEvent.java:10]: What is this override doing here?

File: src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java [hunk 3]
```diff
@@ -987,4 +989,90 @@ public Version since() {
...
+		@Override
+		public String getName() {
+			return "item_pre_anvil";
+		}
+
+		@Override
+		public String docs() {
+			return "{}"
+					+ " Fires when a recipe is formed in an anvil, but the result has not yet been clicked."
+					+ " { viewers: all human entities looking at this anvil "
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java:1005]: I would word this as `all players viewing this anvil's interface` or something similar that is a bit more explicit.

File: src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java [hunk 4]
```diff
@@ -987,4 +989,90 @@ public Version since() {
...
+					+ " | first_item: the first item in the anvil"
+					+ "	| second_item: the second item in the anvil"
+					+ " | result: the result of the anvil"
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java:1008]: I would word this as:
`the first ingredient in the anvil recipy`
`the second ingredient in the anvil recipy`
`the result of the anvil recipy`
Or something alike. Again a bit more explicit, leaving less room for misinterpretation (what is the result of an anvil?).

File: src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java [hunk 5]
```diff
@@ -987,4 +989,90 @@ public Version since() {
...
+		}
+
+		@Override
+		public BindableEvent convert(CArray manualObject, Target t) {
+			throw ConfigRuntimeException.CreateUncatchableException("Unsupported operation.", Target.UNKNOWN);
+		}
+
+		@Override
+		public Map<String, Mixed> evaluate(BindableEvent event) throws EventException {
+			if(event instanceof MCPrepareAnvilEvent) {
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java:1031]: Java now has a nice language feature that you can use here.
```java
if(event instanceof MCPrepareAnvilEvent) {
    MCPrepareAnvilEvent e = (MCPrepareAnvilEvent) event;
```
is equivalent to:
```java
if(event instanceof MCPrepareAnvilEvent e) {
```

It's up to you to decide whether or not you want to use that, but it's been used more and more in CH.

**Review by Pieter12345 — COMMENTED** [review]
Quick review without having compiled or executed the code. It looks like you were successful at sticking to the project structure! Thanks for contributing :)

File: src/main/java/com/laytonsmith/abstraction/bukkit/BukkitMCAnvilInventory.java (hunk 1 above)
Comment (Lildirt) [reply src/main/java/com/laytonsmith/abstraction/bukkit/BukkitMCAnvilInventory.java]: Indeed it should! Updated.

File: src/main/java/com/laytonsmith/abstraction/events/MCPrepareAnvilEvent.java (hunk 2 above)
Comment (Lildirt) [reply src/main/java/com/laytonsmith/abstraction/events/MCPrepareAnvilEvent.java]: Uh, nothing. Actually, that entire `getInventory()` definition doesn't need to exist. It's defined in `MCInventoryEvent`. Removed it.

File: src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java (hunk 3 above)
Comment (Lildirt) [reply src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java]: Verbage updated.

File: src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java (hunk 4 above)
Comment (Lildirt) [reply src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java]: Verbage updated.

File: src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java (hunk 5 above)
Comment (Lildirt) [reply src/main/java/com/laytonsmith/core/events/drivers/InventoryEvents.java]: Sure! I've updated it to use the short-hand.

**Comment by PseudoKnight** [comment]
Looks like you matched 'viewers' from item_pre_craft, whereas we should probably get a player for both of these from the InventoryView. I don't think there can be multiple viewers, at least for these events on these GUIs. I would recommend removing 'viewers' and add 'player' for this event.

There's one hitch. InventoryView was changed from an abstract class to an interface a couple of weeks ago. Directly using methods from that interface may not work on older versions of the server. Might need reflection. Something like:

```
	@Override
	public MCHumanEntity getPlayer() {
		return new BukkitMCHumanEntity(ReflectionUtils.invokeMethod(e.getView(), "getPlayer"));
	}
```

**Comment by PseudoKnight** [comment]
Let me test the other event to see if reflection is needed here too.

**Comment by PseudoKnight** [comment]
Actually can probably just do e.getViewers.get(0) instead of reflection. Or should we have that as a fallback for getPlayer() on older versions?

**Comment by Lildirt** [comment]
> Actually can probably just do e.getViewers.get(0) instead of reflection. That's probably the way to go.

Pushed a commit. Something like that?

**Comment by PseudoKnight** [comment]
That looks fine. It looks impossible to have more than one viewer for an anvil. Unlike shared inventory views, the list of tracked viewers of an anvil is created every time a call to open an anvil is used.

**Comment by PseudoKnight** [comment]
Just noticed you have question marks in the description of the repair costs. Perhaps they were temporary?

**Pull request merged**